theano/gof/type.py

```diff
@@ -795,3 +796,137 @@ def signature(self):
         return (self.type,)
 
 CDataType.Constant = CDataTypeConstant
+
+
+class EnumType(Type, dict):
```

Comment:
There may be some very eccentric solutions if we allow using C++ syntax.

For example, we could use a special constant type which handle only C-defined constants:

```python
# we will need a type (for C code) and a name (to use as class name).
params_type = CEnumType('CUDNN_SOFTMAX_SOFT', 'CUDNN_SOFTMAX_ACCURATE',
              dtype='int32', name='MyAlgo')
```

Then in Python, we will use specific values, different from real C values, but that may be used for mapping steps (e.g array index values):

```python
params_type.CUDNN_SOFTMAX_SOFT # == 0
params_type.CUDNN_SOFTMAX_ACCURATE # == 1
```

We allow `CEnumType` to have values, like the other classic Theano types. Here it will be the `algo` value:
```python
# in __init__
self.algo = self.params_type.CUDNN_SOFTMAX_ACCURATE # == 1

# in get_params() (automatically handled by Wrapper is CEnumType is defined into a Wrapper)
return self.algo
```

But when converted to C++ code, we could create a specific class to represent the value passed in runtime, in which we overload the comparison operators:

```C++
/* PS: I have forgotten C++ syntax a little, but I could recover it quickly ! */

struct MyAlgo {
    static dtype map[2];
    int value;
    MyAlgo(int runtime_value): value(runtime_value) {}
    bool operator==(dtype other_value) {
        // The trick is here.
        return MyAlgo::map[this->value] == other_value;
    }
}
MyAlgo::map = {CUDNN_SOFTMAX_SOFT, CUDNN_SOFTMAX_ACCURATE};
```

Then, in main C function, we could write:
```C++
/* Algo would be of type MyAlgo */
if (algo == CUDNN_SOFTMAX_ACCURATE); // ...
```

A little hard, but if we allow C++ syntax, I think this is possible, although I will have to recover C++ practice. What do you think ?